File: Diagrammatize.Design.Patterns/code/cxx/Observer/A1/NumberGenerator.cpp
#include "NumberGenerator.hpp"
#include "Observer.hpp"

NumberGenerator::~NumberGenerator() {
}

void NumberGenerator::addObserver(Observer* observer) {
    observers.push_back(observer);
}

void NumberGenerator::deleteObserver(Observer* observer) {
    observers.remove(observer);
}

void NumberGenerator::notifyObservers() {
    for (auto o: observers) {
        o->update(this);
    }
}

File: Diagrammatize.Design.Patterns/code/python/Observer/A2/NumberGenerator.py
import abc

class NumberGenerator(abc.ABC):
    def __init__(self):
        self.observers = list()

    def addObserver(self, observer):
        self.observers.append(observer)

    def deleteObserver(self, observer):
        self.observers.remove(observer)

    def notifyObservers(self):
        for o in self.observers:
            o.update(self)

    @abc.abstractmethod
    def getNumber(self):
        pass

    @abc.abstractmethod
    def execute(self):
        pass

